### DISCUSSION TITLE
How can I avoid persisting large Spark dataframes between ops?

### DISCUSSION THREAD

**Discussion by sryza** [dagster-io/dagster#6899] [Q&A] [answered]
This question has come up a few times in #dagster-support. E.g.

> Assuming the DF is petabytes in size I do not neccessarily want to materialize all this IO. Spark itself will create a DAG of the submitted operations - and perhaps calculate additional predicate pushdowns or projections for optimization (AQE). How can I use dagster and ops to define multiple (reusable building blocks) but still not materialize the IO between these steps?

Another example:
> If one of my Ops yields a Dataframe a terabyte in size that’s passed to a subsequent op, is it really preferred to write that to disk only to read it again?

**Comment by sryza** [comment, marked as answer]
### mem_io_manager

To store intermediate values in memory, instead of on-disk, you can use `mem_io_manager`.  For `mem_io_manager` to work, all the ops within an execution need to share the same process, which requires using the `in_process_executor`.

With software-defined assets this looks like:
```python
from dagster import mem_io_manager, in_process_executor, Definitions

defs = Definitions(
    assets=...,
    resource_defs={"io_manager": mem_io_manager}, executor_def=in_process_executor)
)
```


For a job, this looks like:
```python
from dagster import mem_io_manager, in_process_executor

@job(resource_defs={"io_manager": mem_io_manager}, executor_def=in_process_executor)
def my_job():
    ...
```

If you want only some of the outputs of your ops or assets to be stored in memory. You can assign different IO managers to different outputs and assets with `io_manager_key`: https://docs.dagster.io/concepts/io-management/io-managers#per-output-io-manager.

### But do you need multiple ops?

An alternative approach is to move from assets/ops to vanilla Python functions. There are a couple reasons to split logic into multiple ops:
* To enable checkpointing and re-executing from the middle if something fails near the end. If you care about that, then persisting the data to disk is a necessity.
* Making the structure of your DAG visually inspectable in Dagit.

If you don't care about these, then it might make more sense not to split your PySpark transformation across multiple ops, and instead just use vanilla Python functions.

**Reply by zyd14** [reply]
This all makes sense to me - it's a bit of an inherent limitation (or feature?) of spark's in-memory model.  I've been essentially doing the latter (not splitting across ops) - having an op call the vanilla Python code, with an op containing all the calls necessary to those Python functions (distributed in a separately packaged library) in order to perform a full transformation and materialization (i.e. write to disk) of a dataset. You obviously lose the checkpointing and high-granularity DAG visualization, but you can still orchestrate larger individual materializations  as ops in part of a larger pipeline, and get DAG visualization and checkpointing at that level (which is still worth a lot to me).

**Reply by alangenfeld** [reply]
Some related issues that could change the options if addressed:

https://github.com/dagster-io/dagster/issues/6834

https://github.com/dagster-io/dagster/issues/4041

**Comment by emirkmo** [comment]
When is the memory "cleared" when using the in memory IO manager? When the run finishes?

**Reply by sryza** [reply]
Memory isn't shared across runs, but the objects won't be garbage-collected until the result object goes out of scope.

**Comment by zhh210** [comment]
does this solution still work on the latest version? `executor_def` is no longer a valid argument of `Definitions`. It keeps getting me an error of 
```
pyspark.errors.exceptions.base.PySparkRuntimeError: [CONTEXT_ONLY_VALID_ON_DRIVER] It appears that you are attempting to reference SparkContext from a broadcast variable, action, or transformation. SparkContext can only be used on the driver, not in code that it run on workers. For more information, see SPARK-5063.
```   

The full example with failure msg is in this [discussion](https://github.com/dagster-io/dagster/discussions/18210).
